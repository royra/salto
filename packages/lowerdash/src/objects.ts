/*
*                      Copyright 2020 Salto Labs Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with
* the License.  You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
import { inspect } from 'util'
import { DefaultMap } from './collections/map'

export class DuplicateKeysError extends Error {
  constructor(readonly duplicates: Record<string, unknown[]>) {
    super(`Duplicate keys in: ${inspect(duplicates)}`)
  }
}

export const fromPairsEnsureUnique = <V>(pairs: [string, V][]): Record<string, V> => {
  const m = new DefaultMap<string, V[]>(() => [])
  const duplicates: Record<string, V[]> = {}
  const result: Record<string, V> = {}
  for (const [key, value] of pairs) {
    const values = m.get(key)
    if (values.push(value) === 2) {
      duplicates[key] = values
    }
    result[key] = value
  }
  if (Object.keys(duplicates).length > 0) {
    throw new DuplicateKeysError(duplicates)
  }
  return result
}
