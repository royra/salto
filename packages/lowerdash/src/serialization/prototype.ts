/*
*                      Copyright 2020 Salto Labs Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with
* the License.  You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
import { inspect } from 'util'
import { Replacer, Reviver, isPrimitive } from './common'

export const encodePrototypes = (
  prototypes: Map<object, string>,
  prototypeFieldName: string,
): Replacer => (_propName: string, propValue: unknown): unknown => {
  if (isPrimitive(propValue)) {
    return propValue
  }
  const prototypeFieldValue = prototypes.get(Object.getPrototypeOf(propValue))
  if (prototypeFieldValue === undefined) {
    return propValue
  }
  return Object.assign(propValue, { [prototypeFieldName]: prototypeFieldValue })
}

class UnknownPrototypeError extends Error {
  constructor(
    readonly prototypeKey: string,
    readonly object: unknown,
    readonly propName: string,
  ) {
    super(`unknown prototype "${prototypeKey}" while reviving "${propName}" in object ${inspect(object)}`)
  }
}

export const revivePrototypes = (
  prototypes: Map<string, object>,
  prototypeFieldName: string,
): Reviver => {
  function reviver(this: unknown, propName: string, propValue: never): unknown {
    if (isPrimitive(propValue)) {
      return propValue
    }
    if (!(prototypeFieldName in propValue) || typeof propValue[prototypeFieldName] !== 'string') {
      return propValue
    }

    const prototype = prototypes.get(propValue[prototypeFieldName])
    if (prototype === undefined) {
      throw new UnknownPrototypeError(propValue[prototypeFieldName], this, propName)
    }
    return Object.setPrototypeOf(propValue, prototype)
  }

  return reviver as Reviver
}
