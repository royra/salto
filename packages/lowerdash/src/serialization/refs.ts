/*
*                      Copyright 2020 Salto Labs Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with
* the License.  You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
import { DefaultMap } from '../collections/map'
import { Replacer, isPrimitive } from './common'

export type WithRefs = {
  root: unknown
  refs: Record<string, string[]>
}

const NULL_REPLACER: Replacer = (_propName: string, propValue: unknown) => propValue

export type RefPath = string[]

type Node = {
  [key: string]: Node | unknown
}

const navigateTo = (root: Node, path: string): [string, Node] => {
  const parts = path.split(/(?<!\\)\./).map(p => p.replace(/\\\./g, '.'))
  let o = root
  for (let i = 0; i < parts.length - 1; i += 1) {
    o = o[parts[i]] as Node
  }
  return [parts[parts.length - 1], o]
}

const getRef = (root: Node, path: string): unknown => {
  const [propName, object] = navigateTo(root, path)
  return object[propName]
}

const setRef = (root: Node, path: string, value: unknown): void => {
  const [propName, object] = navigateTo(root, path)
  object[propName] = value
}

const serializeRef = (paths: string[]): string => paths
  .map(p => p.replace(/\./g, '\\.'))
  .join('.')

export const refsReplacer = (
  innerReplacer: Replacer = NULL_REPLACER,
): Replacer => {
  const visited = new Map<unknown, RefPath>()
  const refs = new DefaultMap<string, string[]>(() => [])

  const rootResult: Omit<WithRefs, 'refs'> & { refs: DefaultMap<string, string[]> } = { root: undefined, refs }
  let visitedRefs = false

  const addRef = (targetPath: string[], refPath: string[]): void => {
    refs.get(serializeRef(targetPath)).push(serializeRef(refPath))
  }

  function replacer(this: unknown, propName: string, propValue: unknown): unknown {
    // console.log('replacer, this=%o, propName=%o propValue=%o', this, propName, propValue)
    if (visitedRefs) { // somewhere inside refs
      return propValue
    }

    if (rootResult.refs === propValue) { // refs prop
      visitedRefs = true
      return Object.fromEntries(refs.entries())
    }

    const result = innerReplacer.call(this, propName, propValue)

    if (rootResult.root === undefined) { // root object - replace with WithRefs
      rootResult.root = result
      return rootResult
    }

    if (rootResult.root === result) { // no need to process root
      return result
    }

    if (isPrimitive(result)) {
      return result
    }

    const parentRefPath = visited.get(this)
    const refPath = [...parentRefPath ?? [], propName]

    const visitedRefPath = visited.get(result)

    if (visitedRefPath === undefined) {
      visited.set(result, refPath)
      return result
    }

    addRef(visitedRefPath, refPath)
    return undefined
  }

  return replacer
}

export const resolveRefs = (
  { refs, root }: WithRefs,
): unknown => {
  Object.entries(refs).forEach(([targetPathStr, refPaths]) => {
    const target = getRef(root as Node, targetPathStr)
    refPaths.forEach(refPathStr => {
      setRef(root as Node, refPathStr, target)
    })
  })
  return root
}
