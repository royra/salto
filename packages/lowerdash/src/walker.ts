/*
*                      Copyright 2020 Salto Labs Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with
* the License.  You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
import { isPrimitive } from './values'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type WalkCallback<Return> = (this: any, propName: string, propValue: unknown) => Return

// Walk the specified value, executing the callback function exactly once per property.
// The callback is given the property name and value as arguments,
// and the containing object as `this`.
export type Walker = (value: unknown, callback: WalkCallback<void>) => void

// If the callback function returns undefined, the property is deleted from the object.
// Otherwise, the property is redefined to be the return value.
export type MutatingWalker = (value: unknown, callback: WalkCallback<unknown>) => unknown

// Convert Walker to MutatingWalker
export const toMutating = (walk: Walker): Walker => (value, callback) => {
  walk(value, function mutatingWalker(propName, propValue): void {
    const mutation = callback.call(this, propName, propValue)

    if (mutation === undefined) {
      delete this[propName]
    } else {
      this[propName] = mutation
    }
  })

  return value
}

// Walk the specified value, beginning with the most nested properties and proceeding
// to the original value itself.
export const walkBottomUpDepthFirst: Walker = (value, callback) => {
  const visited = new Set<unknown>()

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  function walk(this: any, propName: string, propValue: unknown): void {
    const isPrimitiveProp = isPrimitive(propValue)

    if (!isPrimitiveProp) {
      if (visited.has(propValue)) {
        return
      }
      visited.add(propValue)
    }

    if (Array.isArray(propValue)) {
      propValue.forEach((v, i) => {
        walk.call(propValue, String(i), v)
      })
    } else if (!isPrimitiveProp) {
      Object.getOwnPropertyNames(propValue).forEach(p => {
        walk.call(propValue, p, (propValue as Record<string, unknown>)[p])
      })
    }

    callback.call(this, propName, propValue)
  }

  const topContext = { '': value }
  walk.call(topContext, '', topContext[''])
}

// Modeled after the JSON.parse reviver, described here: https://tc39.es/ecma262/#sec-json.parse
export const mutateBottomUpDepthFirst = toMutating(walkBottomUpDepthFirst)
