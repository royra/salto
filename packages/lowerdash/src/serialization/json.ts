/*
*                      Copyright 2020 Salto Labs Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with
* the License.  You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
import { inspect } from 'util'
import { defaultOpts } from '../functions'
import { isPrimitive } from '../values'
import { mutateBottomUpDepthFirst } from '../walker'
import { getRef, RefNode, RefPath, serializeRef, setRef } from './refs'

type ReplacerResult = {
  data: unknown
  refs: Record<string, string[]> // target => refs
  prototypes: Record<string, string[]> // prototypeName => objects
}

const addInObj = <T>(o: Record<string, T[]>, key: string, value: T): void => {
  let entry = o[key]
  if (!entry) {
    entry = []
    o[key] = entry
  }
  entry.push(value)
}

const resolveRefs = (
  { refs, data }: ReplacerResult,
): void => {
  Object.entries(refs).forEach(([targetPathStr, refPaths]) => {
    const target = getRef(data as RefNode, targetPathStr)
    refPaths.forEach(refPathStr => {
      setRef(data as RefNode, refPathStr, target)
    })
  })
}

export class UnknownPrototypeError extends Error {
  constructor(
    readonly prototypeName: string,
    readonly refPaths: string[],
  ) {
    super(`Unknown prototype "${prototypeName}" at paths: ${inspect(refPaths)}`)
  }
}

const revivePrototypes = (
  knownPrototypes: Map<string, object>,
  { prototypes, data }: ReplacerResult,
): void => {
  Object.entries(prototypes).forEach(([prototypeName, refPaths]) => {
    const prototype = knownPrototypes.get(prototypeName)
    if (prototype === undefined) {
      throw new UnknownPrototypeError(prototypeName, refPaths)
    }

    refPaths.forEach(refPathStr => {
      const target = getRef(data as RefNode, refPathStr)
      Object.setPrototypeOf(target, prototype)
    })
  })
}

const revive = (
  replacerResult: ReplacerResult,
  knownPrototypes: Map<string, object>,
): unknown => {
  revivePrototypes(knownPrototypes, replacerResult)
  resolveRefs(replacerResult)
  return replacerResult.data
}

export type JsonSerializer = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  parse(text: string, reviver?: (this: any, key: string, value: any) => any): any
  stringify(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    value: any, replacer?: (this: any, key: string, value: any) => any, space?: string | number
  ): string
}

export type JsonSerializerOpts = {
  knownPrototypes: [object, string][]
  useSerializer: JsonSerializer
}

export class DuplicatePrototypeValue extends Error {
  constructor(readonly knownPrototypes: [object, string][]) {
    super(`Duplicate prototype in list of known prototypes: ${inspect(knownPrototypes)}`)
  }
}

const protoToName = (knownPrototypes: [object, string][]): Map<object, string> => {
  const result = new Map<object, string>(knownPrototypes)

  if (result.size !== knownPrototypes.length) {
    throw new DuplicatePrototypeValue(knownPrototypes)
  }

  return result
}

export class DuplicatePrototypeName extends Error {
  constructor(readonly knownPrototypes: [object, string][]) {
    super(`Duplicate name in list of known prototypes: ${inspect(knownPrototypes)}`)
  }
}

const nameToProto = (knownPrototypes: [object, string][]): Map<string, object> => {
  const result = new Map<string, object>(
    knownPrototypes.map(([proto, name]) => [name, proto])
  )

  if (result.size !== knownPrototypes.length) {
    throw new DuplicatePrototypeName(knownPrototypes)
  }

  return result
}

const runReviver = mutateBottomUpDepthFirst

const serializer = (
  { knownPrototypes, useSerializer }: JsonSerializerOpts,
): JsonSerializer => ({
  stringify: (value, innerReplacer, space) => {
    const prototypeToName = protoToName(knownPrototypes)
    const visited = new Map<unknown, RefPath>()
    const refs: Record<string, string[]> = {}
    const prototypes: Record<string, string[]> = {}

    const checkPrototype = (propValue: unknown, refPathStr: string): void => {
      const prototypeName = prototypeToName.get(Object.getPrototypeOf(propValue))
      if (prototypeName !== undefined) {
        addInObj(prototypes, prototypeName, refPathStr)
      }
    }

    const addRef = (targetPath: RefPath, refPathStr: string): void => {
      addInObj(refs, serializeRef(targetPath), refPathStr)
    }

    let initialVisit = true

    function wrappedReplacer(this: unknown, propName: string, propValue: unknown): unknown {
      if (innerReplacer) {
        // eslint-disable-next-line no-param-reassign
        propValue = innerReplacer.call(this, propName, propValue)
      }

      if (initialVisit) { // propName ''
        initialVisit = false
        return propValue
      }

      if (isPrimitive(propValue)) {
        return propValue
      }

      const parentRefPath = visited.get(this)
      const refPath = [...parentRefPath ?? [], propName]
      const refPathStr = serializeRef(refPath)

      const visitedRefPath = visited.get(propValue)

      if (visitedRefPath === undefined) { // object not seen before
        visited.set(propValue, refPath)
        checkPrototype(propValue, refPathStr)
        return propValue
      }

      addRef(visitedRefPath, refPathStr)
      return undefined // remove recurring ref
    }

    const serializedData = useSerializer.stringify(value, wrappedReplacer, space)
    const serializedPrototypes = useSerializer.stringify(prototypes, undefined, space)
    const serializedRefs = useSerializer.stringify(refs, undefined, space)

    return `{"prototypes":${serializedPrototypes},"refs":${serializedRefs},"data":${serializedData}}`
  },

  parse: (text, innerReviver) => {
    const revived = revive(
      useSerializer.parse(text) as ReplacerResult,
      nameToProto(knownPrototypes),
    )

    return innerReviver
      ? runReviver(revived, innerReviver)
      : revived
  },
})

/**
 * Drop-in replacement for JSON.serializer and JSON.parse, with the following features:
 * - Prototypes can be preserved, as long as they are known in advance.
 * - References are kept, including circular references.
 *
 * @example Preserving instances
 *
 * class MyClass { x = 12 }
 * const myInstance = new MyClass()
 *
 * const serializer = jsonSerializer({ knownPrototypes: [[MyClass.prototype, 'MyClass']] })
 * const str = serializer.stringify(myInstance)
 * const revivedInstance = serializer.parse(str) as MyClass
 * console.log(revivedInstance instanceof MyInstance) // true
 * console.log(revivedInstance.x) // 12
 *
 * @example Keeping references
 *
 * const owner = { name: 'john' }
 * const dog = { name: 'bud', owner }
 * const cat = { name: 'lucy', owner }
 *
 * const serializer = jsonSerializer()
 * const str = serializer.stringify([ dog, owner, cat ]) // order does not matter
 * const [dDog, dOwner, dCat] = serializer.parse(str) as [typeof dog, typeof owner, typeof cat]
 * console.log(dDog.owner === dCat.owner) // true
 *
 */
//
//
export const jsonSerializer = defaultOpts<JsonSerializerOpts, JsonSerializer>(
  serializer,
  {
    knownPrototypes: [],
    useSerializer: JSON,
  }
)
