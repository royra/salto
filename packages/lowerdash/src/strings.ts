/*
*                      Copyright 2020 Salto Labs Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with
* the License.  You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
import { walkBottomUpDepthFirst } from './walker'

export const DEFAULT_ALPHABET = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'
export const LOWERCASE = 'abcdefghijklmnopqrstuvwxyz'

export const insecureRandomString = (
  { length = 10, alphabet = DEFAULT_ALPHABET }: {
    length?: number
    alphabet?: string
  } = {},
): string => Array(...Array(length))
  .map(() => alphabet.charAt(Math.floor(Math.random() * alphabet.length)))
  .join('')

const collatorOptions = {
  locale: 'en',
  usage: 'sort',
  sensitivity: 'variant',
  ignorePunctuation: false,
  collation: 'default',
  numeric: false,
  caseFirst: 'false',
}

export const stableCollator = Intl.Collator(collatorOptions.locale, collatorOptions)

// This method solves a memory leak which takes place when we use slices
// from a large string in order to populate the strings in the elements.
// v8 will attempt to optimize the slicing operation by internally representing
// the slices string as a pointer to the large string with a start and finish indexes
// for the slice. As a result - the original string will not be evacuated from memory.
// to solve this we need to force v8 to change the sliced string representation to a
// regular string.
export const detachString = (s: string): string => Buffer.from(s).toString()

export const detachStrings = <T>(v: T): T => {
  walkBottomUpDepthFirst(
    v,
    function replacer(propName, propValue) {
      if (
        typeof propValue === 'string'
        /* istanbul ignore next */
        && Object.getOwnPropertyDescriptor(this, propName)?.writable
      ) {
        this[propName] = detachString(propValue)
      }
    },
  )

  return v
}
