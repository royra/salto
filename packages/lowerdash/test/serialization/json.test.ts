/*
*                      Copyright 2020 Salto Labs Ltd.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with
* the License.  You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/
import { serialization } from '../../src'
import { DuplicatePrototypeName, DuplicatePrototypeValue } from '../../src/serialization'

const { jsonSerializer, UnknownPrototypeError } = serialization
type JsonSerializer = serialization.JsonSerializer

describe('jsonSerializer', () => {
  describe('prototype reviving', () => {
    class Base {
      base = 'base value'
    }

    class Derived extends Base {
      derived = 'derived value'
    }

    type HasPrototype = { prototype: object }

    const classes: [HasPrototype, string][] = [
      [Base, 'Base'],
      [Derived, 'Derived'],
    ]

    const protoToName: [object, string][] = classes
      .map(([cls, name]) => [cls.prototype, name])

    type RootType = [Base, Derived, { x: 12; y: Derived }, 17]
    let root: RootType

    beforeEach(() => {
      root = [new Base(), new Derived(), { x: 12, y: new Derived() }, 17]
    })

    const serializer = jsonSerializer({ knownPrototypes: protoToName })

    describe('roundtrip - prototype encode and decode', () => {
      let parsed: typeof root
      beforeEach(() => {
        const s = serializer.stringify([...root])
        parsed = serializer.parse(s) as typeof parsed
      })

      it('should be an exact copy of the input', () => {
        expect(parsed).toEqual(root)
      })

      it('should restore the prototypes', () => {
        const [b, d, { y }] = parsed
        expect(b).toBeInstanceOf(Base)
        expect(b.base).toEqual('base value')
        expect(d).toBeInstanceOf(Base)
        expect(d.base).toEqual('base value')
        expect(d.derived).toEqual('derived value')
        expect(d).toBeInstanceOf(Derived)
        expect(y).toBeInstanceOf(Derived)
      })
    })

    describe('when given a missing prototype map', () => {
      let s: string
      let serializer2: JsonSerializer
      beforeEach(() => {
        s = serializer.stringify([...root])
        serializer2 = jsonSerializer({
          knownPrototypes: protoToName.filter(([_p, name]) => name !== 'Base'),
        })
      })

      const shouldThrow = (): void => { serializer2.parse(s) }

      it('should throw', () => {
        expect(shouldThrow).toThrow(UnknownPrototypeError)
      })

      describe('the thrown error', () => {
        let error: serialization.UnknownPrototypeError
        beforeEach(() => {
          try {
            shouldThrow()
          } catch (e) {
            error = e
          }
          expect(error).toBeDefined()
        })

        it('should contain informational props', () => {
          expect(error.prototypeName).toEqual('Base')
          expect(error.refPaths).toEqual(['0'])
        })

        it('should contain an informational message', () => {
          expect(error.message).toMatch(/Unknown prototype "Base" at paths: \[/)
        })
      })
    })

    describe('when parse is given a duplicate prototype name', () => {
      let s: string
      let serializer2: JsonSerializer
      let specifiedKnownPrototypes: [object, string][]
      beforeEach(() => {
        s = serializer.stringify([...root])
        specifiedKnownPrototypes = [...protoToName, [{}, protoToName[0][1]]]
        serializer2 = jsonSerializer({
          knownPrototypes: specifiedKnownPrototypes,
        })
      })

      const shouldThrow = (): void => { serializer2.parse(s) }

      it('should throw', () => {
        expect(shouldThrow).toThrow(DuplicatePrototypeName)
      })

      describe('the thrown error', () => {
        let error: serialization.DuplicatePrototypeName
        beforeEach(() => {
          try {
            shouldThrow()
          } catch (e) {
            error = e
          }
          expect(error).toBeDefined()
        })

        it('should contain informational props', () => {
          expect(error.knownPrototypes).toEqual(specifiedKnownPrototypes)
        })

        it('should contain an informational message', () => {
          expect(error.message).toMatch(/Duplicate name in list of known prototypes: \[/)
        })
      })
    })

    describe('when stringify is given a duplicate prototype', () => {
      let serializer2: JsonSerializer
      let specifiedKnownPrototypes: [object, string][]
      beforeEach(() => {
        specifiedKnownPrototypes = [...protoToName, [protoToName[0][0], 'x']]
        serializer2 = jsonSerializer({
          knownPrototypes: specifiedKnownPrototypes,
        })
      })

      const shouldThrow = (): void => { serializer2.stringify([...root]) }

      it('should throw', () => {
        expect(shouldThrow).toThrow(DuplicatePrototypeValue)
      })

      describe('the thrown error', () => {
        let error: serialization.DuplicatePrototypeValue
        beforeEach(() => {
          try {
            shouldThrow()
          } catch (e) {
            error = e
          }
          expect(error).toBeDefined()
        })

        it('should contain informational props', () => {
          expect(error.knownPrototypes).toEqual(specifiedKnownPrototypes)
        })

        it('should contain an informational message', () => {
          expect(error.message).toMatch(/Duplicate prototype in list of known prototypes: \[/)
        })
      })
    })
  })

  describe('refs', () => {
    const o1 = { x: 12, y: 13 }
    const o2 = { x: 14, y: 15 }

    const serializer = jsonSerializer()

    const testRoundtrip = (description: string, rootFactory: () => unknown): void => {
      test(`when given ${description}`, () => {
        const root = rootFactory()
        const s = serializer.stringify(root)
        const parsed = serializer.parse(s)
        expect(parsed).toEqual(root)
      })
    }

    testRoundtrip('a single scalar', () => 12)

    testRoundtrip('no duplicate refs', () => {
      const el1 = { o1 }
      const el2 = { o2, z: 16 }
      return [el1, el2]
    })

    testRoundtrip('nulls', () => {
      const el1 = { o1 }
      const el2 = { o2, z: 16, null1: null, listWithNull: ['a', null] }
      return [el1, el2, null]
    })

    testRoundtrip('a duplicate ref', () => {
      const el1 = { o1 }
      const el2 = { o1, o2, z: 16 }
      return [el1, el2]
    })

    testRoundtrip('a deep duplicate ref', () => {
      const el1 = { o1 }
      const el2 = { foo: { bar: { baz: o1 } }, o2, z: 16 }
      return [el1, el2]
    })

    testRoundtrip('a circular ref', () => {
      const el1 = { o1 }
      const el2 = { o1, o2, z: 16 }
      Object.assign(el1, { el2 })
      Object.assign(el2, { el1 })
      return [el1, el2]
    })

    testRoundtrip('a duplicate ref whose name contains a dot', () => {
      const nameWithDot = 'name.dot'
      const el1 = { [nameWithDot]: o1 }
      const el2 = { [nameWithDot]: o1, o2, z: 16 }
      return [el1, el2]
    })

    test('circular refs', () => {
      const el1 = { o1: { ...o1 } as Record<string, unknown> }
      el1.o1.parent = el1
      const root = [el1]
      const s = serializer.stringify(root, undefined, 4)
      const parsed = serializer.parse(s) as typeof root
      expect(parsed[0].o1.parent).toBe(parsed[0])
    })

    describe('when a custom replacer is specified', () => {
      const el1 = { o1 }
      const el2 = { o1, o2, z: 16 }
      const el3 = { o1, o2, z: 16 }
      const root = [el1, el2, el3]

      function customReplacer(this: unknown, k: string, v: unknown): unknown {
        if (this === el2 && k === 'o1' && v === o1) {
          return 17
        }
        return v
      }

      let parsed: unknown

      beforeEach(() => {
        const s = serializer.stringify(root, customReplacer)
        parsed = serializer.parse(s)
      })

      it('should call it correctly', () => {
        expect(parsed).toEqual([
          el1,
          { o1: 17, o2, z: 16 },
          el3,
        ])
      })
    })

    describe('when a custom reviver is specified', () => {
      const el1 = { el1o1: o1 }
      const el2 = { el2o1: o1, el2o2: o2, el2z: 16 }
      const el3 = { el3o1: { ...o1 } as Record<string, unknown>, el3o2: o2, el3z: 16 }
      el3.el3o1.parent = el3
      const root = [el1, el2, el3]
      const calls: [unknown, unknown, unknown][] = []

      function customReviver(this: unknown, k: string, v: unknown): unknown {
        calls.push([this, k, v])
        if (k === 'el1o1') {
          expect(this).toEqual({ el1o1: o1 })
          expect(v).toEqual(o1)
          return ['a', 'b']
        }
        if (k === 'el3z') {
          const expectedThis = { el3o1: { ...o1 } as Record<string, unknown>, el3o2: o2, el3z: 16 }
          expectedThis.el3o1.parent = expectedThis
          expect(this).toEqual(expectedThis)
          expect(v).toEqual(16)
          return undefined
        }
        if (v === 'b') {
          throw new Error('should not have called with "b"')
        }
        return v
      }

      let parsed: unknown

      beforeEach(() => {
        const s = serializer.stringify(root)
        parsed = serializer.parse(s, customReviver)
      })

      it('should call it correctly', () => {
        expect(calls).toMatchSnapshot()
      })

      it('should modify the result', () => {
        expect(parsed).toMatchSnapshot()
      })
    })
  })
})
