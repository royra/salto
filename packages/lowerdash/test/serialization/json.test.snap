// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsonSerializer refs when a custom reviver is specified should call it correctly 1`] = `
Array [
  Array [
    Object {
      "x": 12,
      "y": 13,
    },
    "x",
    12,
  ],
  Array [
    Object {
      "x": 12,
      "y": 13,
    },
    "y",
    13,
  ],
  Array [
    Object {
      "el1o1": Array [
        "a",
        "b",
      ],
    },
    "el1o1",
    Object {
      "x": 12,
      "y": 13,
    },
  ],
  Array [
    Array [
      Object {
        "el1o1": Array [
          "a",
          "b",
        ],
      },
      Object {
        "el2o1": Object {
          "x": 12,
          "y": 13,
        },
        "el2o2": Object {
          "x": 14,
          "y": 15,
        },
        "el2z": 16,
      },
      Object {
        "el3o1": Object {
          "parent": [Circular],
          "x": 12,
          "y": 13,
        },
        "el3o2": Object {
          "x": 14,
          "y": 15,
        },
      },
    ],
    "0",
    Object {
      "el1o1": Array [
        "a",
        "b",
      ],
    },
  ],
  Array [
    Object {
      "x": 14,
      "y": 15,
    },
    "x",
    14,
  ],
  Array [
    Object {
      "x": 14,
      "y": 15,
    },
    "y",
    15,
  ],
  Array [
    Object {
      "el2o1": Object {
        "x": 12,
        "y": 13,
      },
      "el2o2": Object {
        "x": 14,
        "y": 15,
      },
      "el2z": 16,
    },
    "el2o2",
    Object {
      "x": 14,
      "y": 15,
    },
  ],
  Array [
    Object {
      "el2o1": Object {
        "x": 12,
        "y": 13,
      },
      "el2o2": Object {
        "x": 14,
        "y": 15,
      },
      "el2z": 16,
    },
    "el2z",
    16,
  ],
  Array [
    Array [
      Object {
        "el1o1": Array [
          "a",
          "b",
        ],
      },
      Object {
        "el2o1": Object {
          "x": 12,
          "y": 13,
        },
        "el2o2": Object {
          "x": 14,
          "y": 15,
        },
        "el2z": 16,
      },
      Object {
        "el3o1": Object {
          "parent": [Circular],
          "x": 12,
          "y": 13,
        },
        "el3o2": Object {
          "x": 14,
          "y": 15,
        },
      },
    ],
    "1",
    Object {
      "el2o1": Object {
        "x": 12,
        "y": 13,
      },
      "el2o2": Object {
        "x": 14,
        "y": 15,
      },
      "el2z": 16,
    },
  ],
  Array [
    Object {
      "parent": Object {
        "el3o1": [Circular],
        "el3o2": Object {
          "x": 14,
          "y": 15,
        },
      },
      "x": 12,
      "y": 13,
    },
    "x",
    12,
  ],
  Array [
    Object {
      "parent": Object {
        "el3o1": [Circular],
        "el3o2": Object {
          "x": 14,
          "y": 15,
        },
      },
      "x": 12,
      "y": 13,
    },
    "y",
    13,
  ],
  Array [
    Object {
      "el3o1": Object {
        "parent": [Circular],
        "x": 12,
        "y": 13,
      },
      "el3o2": Object {
        "x": 14,
        "y": 15,
      },
    },
    "el3o1",
    Object {
      "parent": Object {
        "el3o1": [Circular],
        "el3o2": Object {
          "x": 14,
          "y": 15,
        },
      },
      "x": 12,
      "y": 13,
    },
  ],
  Array [
    Object {
      "el3o1": Object {
        "parent": [Circular],
        "x": 12,
        "y": 13,
      },
      "el3o2": Object {
        "x": 14,
        "y": 15,
      },
    },
    "el3z",
    16,
  ],
  Array [
    Array [
      Object {
        "el1o1": Array [
          "a",
          "b",
        ],
      },
      Object {
        "el2o1": Object {
          "x": 12,
          "y": 13,
        },
        "el2o2": Object {
          "x": 14,
          "y": 15,
        },
        "el2z": 16,
      },
      Object {
        "el3o1": Object {
          "parent": [Circular],
          "x": 12,
          "y": 13,
        },
        "el3o2": Object {
          "x": 14,
          "y": 15,
        },
      },
    ],
    "2",
    Object {
      "el3o1": Object {
        "parent": [Circular],
        "x": 12,
        "y": 13,
      },
      "el3o2": Object {
        "x": 14,
        "y": 15,
      },
    },
  ],
  Array [
    Object {
      "": Array [
        Object {
          "el1o1": Array [
            "a",
            "b",
          ],
        },
        Object {
          "el2o1": Object {
            "x": 12,
            "y": 13,
          },
          "el2o2": Object {
            "x": 14,
            "y": 15,
          },
          "el2z": 16,
        },
        Object {
          "el3o1": Object {
            "parent": [Circular],
            "x": 12,
            "y": 13,
          },
          "el3o2": Object {
            "x": 14,
            "y": 15,
          },
        },
      ],
    },
    "",
    Array [
      Object {
        "el1o1": Array [
          "a",
          "b",
        ],
      },
      Object {
        "el2o1": Object {
          "x": 12,
          "y": 13,
        },
        "el2o2": Object {
          "x": 14,
          "y": 15,
        },
        "el2z": 16,
      },
      Object {
        "el3o1": Object {
          "parent": [Circular],
          "x": 12,
          "y": 13,
        },
        "el3o2": Object {
          "x": 14,
          "y": 15,
        },
      },
    ],
  ],
]
`;

exports[`jsonSerializer refs when a custom reviver is specified should modify the result 1`] = `
Array [
  Object {
    "el1o1": Array [
      "a",
      "b",
    ],
  },
  Object {
    "el2o1": Object {
      "x": 12,
      "y": 13,
    },
    "el2o2": Object {
      "x": 14,
      "y": 15,
    },
    "el2z": 16,
  },
  Object {
    "el3o1": Object {
      "parent": [Circular],
      "x": 12,
      "y": 13,
    },
    "el3o2": Object {
      "x": 14,
      "y": 15,
    },
  },
]
`;
